src/Sets/LineSegment/intersection.jl

```diff
@@ -34,24 +34,31 @@ function intersection(LS1::LineSegment, LS2::LineSegment)
     m = intersection(L1, L2)
     N = promote_type(eltype(LS1), eltype(LS2))
     if m == L1
-        # determine which segment is in both
-        p1 = max(min(LS1.p[1], LS1.q[1]), min(LS2.p[1], LS2.q[1]))
-        p2 = max(min(LS1.p[2], LS1.q[2]), min(LS2.p[2], LS2.q[2]))
-        q1 = min(max(LS1.p[1], LS1.q[1]), max(LS2.p[1], LS2.q[1]))
-        q2 = min(max(LS1.p[2], LS1.q[2]), max(LS2.p[2], LS2.q[2]))
-        if _isapprox(p1, q1) && _isapprox(p2, q2)
-            return Singleton([p1, p2])  # edges have a point in common
-
-        elseif _leq(p1, q1) && _leq(p2, q2)
-            return LineSegment([p1, p2], [q1, q2])
-
-        else
-            return EmptySet{N}(2)  # no intersection
+        # simple solution: compute the box approximation and find the right vertices (there are only 4)
+        B1 = box_approximation(LS1)
+        B2 = box_approximation(LS2)
+        B = intersection(B1, B2)
+        if isempty(B)
+            return B
         end
-
+        p = nothing
+        vlist = [LS1.p, LS1.q, LS2.p, LS2.q]
+        for v1 in vertices(B)
+            for v2 in vlist
+                if _isapprox(v1, v2)
+                    if isnothing(p)
+                        p = v1
+                    elseif _isapprox(v1, p)
+                        return Singleton(p)
+                    else
+                        return LineSegment(p, v1)
+                    end
+                end
+            end
+        end
+        error("this should not happen")
```

Comment:
Would this work?
```julia
points = [LS1.p, LS1.q, LS2.p, LS2.q]
sorted_points = sort(points, by = p -> (p[1], p[2]))
mid_point1 = sorted_points[2]
mid_point2 = sorted_points[3]

if LazySets._isapprox(mid_point1, mid_point2)
    return Singleton(mid_point1)  # edges have a point in common
else
    return LineSegment(mid_point1, mid_point2)
end
```

Reply:
Yes, and it is much better! I just added a commit use to your implementation.

```julia
julia> @time intersection(s10, s11)  # old version
  0.000021 seconds (45 allocations: 2.672 KiB)
LineSegment{Float64, Vector{Float64}}([0.0, 1.0], [1.0, 0.0])

julia> @time intersection(s10, s11)  # new version
  0.000012 seconds (10 allocations: 480 bytes)
LineSegment{Float64, Vector{Float64}}([0.0, 1.0], [1.0, 0.0])
```

Reply:
Ah I meant only for the lines 44-59 in your previous approach. I haven't found something nicely for the empty set one yet.

Reply:
True. I added efficient code for the box approximation that does not allocate.